### ISSUE TITLE
Annoying fullscreen differences between DirectX and OpenGL

### ISSUE THREAD

**Issue by Thraka** [MonoGame/MonoGame#5530]
I'm using MonoGame 3.6 and `GraphicsDeviceManager.HardwareModeSwitch = false;` and having a hard time working out my code for how I want my game to render. Depending on if I'm using DirectX or OpenGL, I get different behaviors.

| Library | Mode | GraphicsDevice.Viewport.Bounds |
| ------- | ----- | ---------------------|
| DirectX | Original window | 640x400 |
| DirectX | To full | 640x400 |
| DirectX | Back to window | 624x361 |
| OpenGL | Original window | 640x400 |
| OpenGL | To full | 1920x1080 |
| OpenGL | Back to window | 640x400 |

As you can see, DirectX misbehaves when coming out of fullscreen, and OpenGL misbehaves when going to fullscreen.

**Comment by KonajuGames** [comment]
Thanks for these bug reports. Simple repro steps will make this easier to
track down and add unit tests so it doesn't break again.

**Comment by willmotil** [comment]
In the january version im still using any changing of size on the backbuffer from the initial start to a large windows size or to a smaller size causes a huge drop in fps my thought is scaleing but i haven't tested whats actually going on with the backbuffer.

**Comment by aienabled** [comment]
This issue seems to be caused by WinFormsGameWindow no triggering ClientSizeChanged event on window state change.. just reported #5533.

**Comment by mrhelmut** [comment]
Regarding ```HardwareModeSwitch = false``` and OpenGL, I may be wrong, but I believe that there's nothing to do about it, that's just how borderless OpenGL works with SDL2 and I don't know if there's a workaround. Maybe @cra0zy knows this better, I'm not sure.

**Comment by cra0zy** [comment]
OpenGL side is working how it is supposed to.

**Comment by Jjagg** [comment]
> This issue seems to be caused by WinFormsGameWindow no triggering ClientSizeChanged event on window state change.. just reported #5533.

That event is not used internally to respond to size changes, so it's not related.

**Comment by Thraka** [comment]
Shouldn't the viewport always stay the same no matter what? Otherwise it will change how much of the scene is being pulled into and rendered on the backbuffer.

**Comment by Jjagg** [comment]
You mean that MG would handle the scaling for you in soft fullscreen? I guess we need to pick either DX or GL behavior and implement that for all platforms. DX behavior is way more convenient for the user because they don't have to handle resolution independence themselves when using soft full screen. And really IMO the soft full screen mode doesn't add much if we don't handle scaling. That said, there's multiple way to handle upscaling the back buffer image and there's a choice to stretch or maintain aspect ratio. Also I don't know if UWP has support for something like this. Users could still get behavior like current DesktopGL by just resizing the back buffer themselves to the display resolution before turning on soft fullscreen.

> Regarding HardwareModeSwitch = false and OpenGL, I may be wrong, but I believe that there's nothing to do about it, that's just how borderless OpenGL works with SDL2 and I don't know if there's a workaround.

If we were to go the DX way we can use https://wiki.libsdl.org/SDL_RenderSetLogicalSize for DesktopGL. That does letter/pillarboxing and I think linear filtering, which is IMO the best way to do this if MG will handle scaling. I guess that's what DX does too?

**Comment by Jjagg** [comment]
> DirectX 	Back to window 	624x361

I'm thinking this might be because of a window bounds/client bounds mess up.

**Comment by cra0zy** [comment]
I would rather use the current OpenGL way of soft fullscreen than the DirectX way, if the programmer wants they can just overwrite the viewport anyway...

**Comment by Thraka** [comment]
And to clarify why I see issues more than perhaps the normal user is that my engine is pretty pixel-based so any distortion to viewport/resolution really messes things up. This is why I went with soft fullscreen because it (most likely) keeps the monitor at the native res and makes pixel detail look nice.

Either way, as long as it's consistent across platforms, we can code any workaround we need for the desired behavior. 😄

**Comment by cra0zy** [comment]
@Thraka You should look at using a RenderTarget2D for your engine.

**Comment by Thraka** [comment]
I am.

**Comment by mrhelmut** [comment]
From my point of view, the current GL behavior is what appears to be the most compatible one across platforms, especially when Retina Macs are involved. I'm pretty sure that having a constant framebuffer size in borderless mode would totally break on a Retina screen.

**Comment by Thraka** [comment]
@mrhelmut I may be confused, but I thought we were talking about viewport, not the buffer, unless I missed something :)

**Comment by Thraka** [comment]
Some more investigation. In DX when I go full screen, then back, move the window with the mouse, then do full screen, then back, over and over, you can see the window actually getting smaller and smaller each time.

**Comment by willmotil** [comment]
Let me just point out a couple of things.

Going to full screen with a small back buffer means your game will look like its pixelated and blocky.

Whats the point of setting preferred ? ... not a question im asking myself.

Haveing the back buffer larger then the size of the window in windows mode is probably a waste but the programer might want it that way i suppose its free anti aliasing since your gpu is forced to down sample the entire thing. 
If he does want to control it he should be handling that with prefered presentation parameters as he is defining or has defined his preferences and monogame should do its best to give him what he asks for.

If no requests are made then the back buffer should be the size that makes the most sense.
A windowed buffer the size of the window a fullscreen the size of his default monitors fullscreen mode native resolution.

**Comment by Jjagg** [comment]
I disagree. I don't think it's confusing what XNA does at all. It resizes the window to match the backbuffer when back buffer size changes, it resizes the back buffer to match the window when the window size changes and it resizes the viewport to match the back buffer when the back buffer size changes. There's no way to have a back buffer and window of a different size in XNA.

@Thraka that's why this became a discussion about the back buffer size. The viewport is directly affecter by it. So the question is if we make the back buffer size be the active display size when we go into soft full screen or if we leave it at the preferred width/height and use tricks to upscale/downscale it (which might not be possible on all platforms).

**Comment by mrhelmut** [comment]
I believe the 1:1 relationship between bounds and backbuffer is there for a reason: compatibility, which is a topic even more relevant in MonoGame because of all the additional platforms available. Some platforms really don't like that this relationship is different.

What about manual resizing of the window in DX? (```AllowUserResizing```) Does it update the backbuffer?

**Comment by Thraka** [comment]
Gotcha. Personally I don't care how it works as long as the following is met

1. It's consistent across DirectX and OpenGL
2. The programmer can still (after everything auto adjusts) change the viewport.
3. Single code path trigger. (OnBackBufferChanged? OnWindowResize? whatever, I would opt for a backbuffer related name because what if I want to host in a form control or something? It's no longer a window)

This way, however it behaves, I can code around it and not worry that it suddenly changes under me, like it did with 3.5->3.6. If it's set of documented behaviors and rules, it will then be something that can be tested for and thus ensure all platforms/drawing libraries pass.

**Comment by cra0zy** [comment]
I'm gonna add that I would rather have the directx soft fullscreen behavior, I've been thinking of GL viewport functionality, not XNA.
